File: example/src/main/java/com/thesurix/example/gesturerecycler/fragment/EmptyViewFragment.java
package com.thesurix.example.gesturerecycler.fragment;

import com.thesurix.example.gesturerecycler.R;
import com.thesurix.example.gesturerecycler.adapter.MonthsAdapter;
import com.thesurix.example.gesturerecycler.callback.MonthDiffCallback;
import com.thesurix.example.gesturerecycler.model.Month;
import com.thesurix.example.gesturerecycler.model.MonthItem;
import com.thesurix.gesturerecycler.EmptyViewVisibilityListener;
import com.thesurix.gesturerecycler.GestureAdapter;
import com.thesurix.gesturerecycler.GestureManager;

import android.animation.FloatEvaluator;
import android.animation.ValueAnimator;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class EmptyViewFragment extends BaseFragment {

    private MonthsAdapter mAdapter;

    @Nullable
    @Override
    public View onCreateView(final LayoutInflater inflater, final ViewGroup container, final Bundle savedInstanceState) {
        final View rootView = inflater.inflate(R.layout.fragment_recycler, container, false);
        mRecyclerView = rootView.findViewById(R.id.recycler_view);
        return rootView;
    }

    @Override
    public void onViewCreated(final View view, @Nullable final Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        final LinearLayoutManager manager = new LinearLayoutManager(getContext());
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(manager);

        mAdapter = new MonthsAdapter(getContext(), R.layout.linear_item_with_background);
        mAdapter.setData(getMonths());
        mAdapter.setUndoSize(2);
        mAdapter.setDataChangeListener(new GestureAdapter.OnDataChangeListener<MonthItem>() {
            @Override
            public void onItemRemoved(final MonthItem item, final int position) {
                final Snackbar undoSnack = Snackbar.make(view, "Month removed from position " + position, Snackbar.LENGTH_SHORT);
                undoSnack.setAction(R.string.undo_text, new View.OnClickListener() {
                    @Override
                    public void onClick(final View v) {
                        mAdapter.undoLast();
                    }
                });
                undoSnack.show();
            }

            @Override
            public void onItemReorder(final MonthItem item, final int fromPos, final int toPos) {
                final Snackbar undoSnack = Snackbar.make(view, "Month moved from position " + fromPos + " to " + toPos, Snackbar.LENGTH_SHORT);
                undoSnack.setAction(R.string.undo_text, new View.OnClickListener() {
                    @Override
                    public void onClick(final View v) {
                        mAdapter.undoLast();
                    }
                });
                undoSnack.show();
            }
        });

        final View emptyView = view.findViewById(R.id.empty_root);
        mAdapter.setEmptyViewVisibilityListener(new EmptyViewVisibilityListener() {
            @Override
            public void onVisibilityChanged(final boolean visible) {
                if (visible) {
                    emptyView.setVisibility(View.VISIBLE);
                    runFadeInAnimation(emptyView);
                } else {
                    runFadeOutAnimation(emptyView);
                }
            }
        });
        mRecyclerView.setAdapter(mAdapter);

        mGestureManager = new GestureManager.Builder(mRecyclerView)
                .setSwipeEnabled(true)
                .setSwipeFlags(ItemTouchHelper.LEFT)
                .setLongPressDragEnabled(true)
                .build();
    }

    @Override
    public void onCreateOptionsMenu(final Menu menu, final MenuInflater inflater) {
        inflater.inflate(R.menu.recycler_empty_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(final MenuItem item) {
        switch (item.getItemId()) {
            case R.id.recycler_undo_menu:
                mAdapter.undoLast();
                break;
            case R.id.recycler_clear_menu:
                mAdapter.clearData();
                break;
            case R.id.recycler_generate_menu:
                final List<MonthItem> months = getMonths();
                final int month = (int)(Math.random() * months.size());
                mAdapter.insert(months.get(month), 0);
                mRecyclerView.scrollToPosition(0);
                break;
            case R.id.recycler_diff_menu:
                final List<MonthItem> diffMonths = getMonths();
                Collections.shuffle(diffMonths);
                mAdapter.setData(diffMonths, new MonthDiffCallback(mAdapter.getData(), diffMonths));
                mRecyclerView.scrollToPosition(0);
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected List<MonthItem> getMonths() {
        final List<MonthItem> monthList = new ArrayList<>();
        monthList.add(new Month("JAN", R.drawable.january));
        monthList.add(new Month("FEB", R.drawable.february));
        monthList.add(new Month("MAR", R.drawable.march));
        monthList.add(new Month("APR", R.drawable.april));
        monthList.add(new Month("MAY", R.drawable.may));
        monthList.add(new Month("JUN", R.drawable.june));

        return monthList;
    }

    private void runFadeInAnimation(final View emptyView) {
        final ValueAnimator fadeInAnimation = ValueAnimator.ofObject(new FloatEvaluator(), 0f, 1f);
        fadeInAnimation.setDuration(getResources().getInteger(R.integer.animation_speed_ms));
        fadeInAnimation.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(final ValueAnimator animation) {
                float alpha = (float) animation.getAnimatedValue();
                emptyView.setAlpha(alpha);
            }
        });
        fadeInAnimation.start();
    }

    private void runFadeOutAnimation(final View emptyView) {
        final ValueAnimator fadeOutAnimation = ValueAnimator.ofObject(new FloatEvaluator(), 1f, 0f);
        fadeOutAnimation.setDuration(getResources().getInteger(R.integer.animation_speed_ms));
        fadeOutAnimation.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(final ValueAnimator animation) {
                float alpha = (float) animation.getAnimatedValue();
                emptyView.setAlpha(alpha);

                if (alpha < 0.01f) {
                    emptyView.setVisibility(View.GONE);
                }
            }
        });

        fadeOutAnimation.start();
    }
}

File: example/src/main/java/com/thesurix/example/gesturerecycler/fragment/EmptyViewFragment.kt
package com.thesurix.example.gesturerecycler.fragment

import android.animation.FloatEvaluator
import android.animation.ValueAnimator
import android.os.Bundle
import android.view.*
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.thesurix.example.gesturerecycler.R
import com.thesurix.example.gesturerecycler.adapter.MonthsAdapter
import com.thesurix.example.gesturerecycler.callback.MonthDiffCallback
import com.thesurix.example.gesturerecycler.model.Month
import com.thesurix.example.gesturerecycler.model.MonthItem
import com.thesurix.gesturerecycler.EmptyViewVisibilityListener
import com.thesurix.gesturerecycler.GestureAdapter
import com.thesurix.gesturerecycler.GestureManager

class EmptyViewFragment : BaseFragment() {

    private var adapter: MonthsAdapter? = null

    private var headerFooterState = false
    override val months: MutableList<MonthItem>
        get() {
            return mutableListOf(Month("JAN", R.drawable.january),
                    Month("FEB", R.drawable.february),
                    Month("MAR", R.drawable.march),
                    Month("APR", R.drawable.april),
                    Month("MAY", R.drawable.may),
                    Month("JUN", R.drawable.june))
        }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_recycler, container, false)
        recyclerView = rootView.findViewById(R.id.recycler_view)
        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val manager = LinearLayoutManager(context)
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = manager

        adapter = MonthsAdapter(R.layout.linear_item_with_background).apply {
            data = months
            setUndoSize(2)
            setDataChangeListener(object : GestureAdapter.OnDataChangeListener<MonthItem> {
                override fun onItemRemoved(item: MonthItem, position: Int, direction: Int) {
                    val undoSnack = Snackbar.make(view, "Month removed from position $position", Snackbar.LENGTH_SHORT)
                    undoSnack.setAction(R.string.undo_text) { adapter?.undoLast() }
                    undoSnack.show()
                }

                override fun onItemReorder(item: MonthItem, fromPos: Int, toPos: Int) {
                    val undoSnack = Snackbar.make(view, "Month moved from position $fromPos to $toPos", Snackbar.LENGTH_SHORT)
                    undoSnack.setAction(R.string.undo_text) { adapter?.undoLast() }
                    undoSnack.show()
                }
            })

            val emptyView = view.findViewById<View>(R.id.empty_root)
            setEmptyViewVisibilityListener(object : EmptyViewVisibilityListener {
                override fun onVisibilityChanged(visible: Boolean) {
                    if (visible) {
                        emptyView.visibility = View.VISIBLE
                        runFadeInAnimation(emptyView)
                    } else {
                        runFadeOutAnimation(emptyView)
                    }
                }
            })
        }

        recyclerView.adapter = adapter

        gestureManager = GestureManager.Builder(recyclerView)
                .setSwipeEnabled(true)
                .setSwipeFlags(ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT)
                .setLongPressDragEnabled(true)
                .setHeaderEnabled(headerFooterState)
                .setFooterEnabled(headerFooterState)
                .build()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.recycler_empty_menu, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.recycler_undo_menu -> adapter?.undoLast()
            R.id.recycler_clear_menu -> adapter?.clearData()
            R.id.recycler_generate_menu -> {
                val months = months
                val month = (Math.random() * months.size).toInt()
                adapter?.insert(months[month], 0)
                recyclerView.scrollToPosition(0)
            }
            R.id.recycler_diff_menu -> {
                val diffMonths = months
                diffMonths.shuffle()
                adapter?.setData(diffMonths, MonthDiffCallback(adapter!!.data, diffMonths))
                recyclerView.scrollToPosition(0)
            }
            R.id.add -> {
                val months = months
                val month = (Math.random() * months.size).toInt()
                adapter?.add(months[month])
            }
            R.id.toggle_hf -> {
                headerFooterState = !headerFooterState
                adapter?.setHeaderEnabled(headerFooterState)
                adapter?.setFooterEnabled(headerFooterState)
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun runFadeInAnimation(emptyView: View) {
        val fadeInAnimation = ValueAnimator.ofObject(FloatEvaluator(), 0f, 1f)
        fadeInAnimation.duration = resources.getInteger(R.integer.animation_speed_ms).toLong()
        fadeInAnimation.addUpdateListener { animation ->
            val alpha = animation.animatedValue as Float
            emptyView.alpha = alpha
        }
        fadeInAnimation.start()
    }

    private fun runFadeOutAnimation(emptyView: View) {
        val fadeOutAnimation = ValueAnimator.ofObject(FloatEvaluator(), 1f, 0f)
        fadeOutAnimation.duration = resources.getInteger(R.integer.animation_speed_ms).toLong()
        fadeOutAnimation.addUpdateListener { animation ->
            val alpha = animation.animatedValue as Float
            emptyView.alpha = alpha

            if (alpha < 0.01f) {
                emptyView.visibility = View.GONE
            }
        }

        fadeOutAnimation.start()
    }
}
